guides/source/active_storage_overview.md

```diff
@@ -0,0 +1,534 @@
+    auth_uri: "https://accounts.google.com/o/oauth2/auth",
+    token_uri: "https://accounts.google.com/o/oauth2/token",
+    auth_provider_x509_cert_url: "https://www.googleapis.com/oauth2/v1/certs",
+    client_x509_cert_url: ""
+  }
+  project: ""
+  bucket: ""
+```
+
+Also, add the Google Cloud Storage client gem to your Gemfile:
+
+``` ruby
+gem "google-cloud-storage", "~> 1.3", require: false
+```
+
+### Mirror Service
+
+You can keep multiple services in sync by defining a mirror service. When
+a file is uploaded or deleted, it's done across all the mirrored services.
+Define each of the services you'd like to use as described above and reference
+them from a mirrored service.
+
+``` yaml
+s3_west_coast:
+  service: S3
+  access_key_id: ""
+  secret_access_key: ""
+  region: ""
+  bucket: ""
+
+s3_east_coast:
+  service: S3
+  access_key_id: ""
+  secret_access_key: ""
+  region: ""
+  bucket: ""
+
+production:
+  service: Mirror
+  primary: s3_east_coast
+  mirrors:
+    - s3_west_coast
+```
+
+NOTE: Files are served from the primary service.
+
+Mirrored services can be used to facilitate a migration between services in
+production. You can start mirroring to the new service, copy existing files from
+the old service to the new, then go all-in on the new service.
+
+If you wish to transform your images, add `mini_magick` to your Gemfile:
+
+``` ruby
+gem 'mini_magick'
+```
+
+Attach Files to a Model
+--------------------------
+One or more files can be attached to a model.
+
+One attachment:
+
+```ruby
+class User < ApplicationRecord
+  # Associates an attachment and a blob. When the user is destroyed they are
+  # purged by default (models destroyed, and resource files deleted).
+  has_one_attached :avatar
+end
+
+# Attach an avatar to the user.
+user.avatar.attach(io: File.open("/path/to/face.jpg"), filename: "face.jpg", content_type: "image/jpg")
+
+class AvatarsController < ApplicationController
+  def update
+    # params[:avatar] contains a ActionDispatch::Http::UploadedFile object
+    Current.user.avatar.attach(params.require(:avatar))
+    redirect_to Current.user
+  end
+end
+```
+
+Many attachments:
+
+```ruby
+class Message < ApplicationRecord
+  has_many_attached :images
+end
+```
+
+```erb
+<%= form_with model: @message, local: true do |form| %>
+  <%= form.text_field :title, placeholder: "Title" %><br>
+  <%= form.text_area :content %><br><br>
+
+  <%= form.file_field :images, multiple: true %><br>
+  <%= form.submit %>
+<% end %>
+```
+
+```ruby
+class MessagesController < ApplicationController
+  def index
+    # Use the built-in with_attached_images scope to avoid N+1
+    @messages = Message.all.with_attached_images
+  end
+
+  def create
+    message = Message.create!(message_params)
+    redirect_to message
+  end
+
+  def show
+    @message = Message.find(params[:id])
+  end
+
+  private
+    def message_params
+      params.require(:message).permit(:title, :content, images: [])
+    end
+end
+```
```

Comment:
We can afford to expand on this, given that it’s the central feature of Active Storage:

> # Attach files to records
>
> ## `has_one_attached`
>
> The `has_one_attached` macro sets up a one-to-one mapping between records and files. Each record can have one file attached to it.
>
> For example, suppose your application has a `User` model. If you want each user to have an avatar, define the `User` model like this:
>
> ```ruby
> class User < ApplicationRecord
>   has_one_attached :avatar
> end
> ```
>
> You can create a user with an avatar:
>
> ```ruby
> class SignupController < ApplicationController
>   def create
>     user = Users.create!(user_params)
>     session[:user_id] = user.id
>     redirect_to root_path
>   end
>
>   private
>     def user_params
>       params.require(:user).permit(:email_address, :password, :avatar)
>     end
> end
> ```
>
> Call `avatar.attach` to attach an avatar to an existing user:
>
> ```ruby
> Current.user.avatar.attach(params[:avatar])
> ```
>
> Call `avatar.attached?` to determine whether a particular user has an avatar:
>
> ```ruby
> Current.user.avatar.attached?
> ```
>
> ## `has_many_attached`
>
> The `has_many_attached` macro sets up a one-to-many relationship between records and files. Each record can have many files attached to it.
>
> For example, suppose your application has a `Message` model. If you want each message to have many images, define the `Message` model like this:
>
> ```ruby
> class Message < ApplicationRecord
>   has_many_attached :images
> end
> ```
>
> You can create a message with images:
>
> ```ruby
> class MessagesController < ApplicationController
>   def create
>     message = Message.create!(message_params)
>     redirect_to message
>   end
>
>   private
>     def message_params
>       params.require(:message).permit(:title, :content, images: [])
>     end
> end
> ```
>
> Call `images.attach` to add new images to an existing message:
>
> ```ruby
> @message.images.attach(params[:images])
> ```
>
> Call `images.attached?` to determine whether a particular message has any images:
>
> ```ruby
> @message.images.attached?
> ```